### PR TITLE
feat: add sha-256 support

### PR DESCRIPTION
**Pull request by gssbzn**
This is an initial approach to add SHA-256 support, is a bit naive on approach to try to be the least destructive of existing methods but initial testing seems like it's working

I'm also adding a Makefile for connivance

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,25 @@
+# A Self-Documenting Makefile: http://marmelab.com/blog/2016/02/29/auto-documented-makefile.html
+
+SOURCE_FILES?=./...
+COVERAGE=coverage.out
+
+export PATH := ./bin:$(PATH)
+
+.PHONY: fmt
+fmt: ## Format code
```
Comment (edaniels) [inline Makefile:9]: I feel like most people's editors have a gofmt builtin. I'd rather exclude this

File: Makefile [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+	@echo "==> Formatting all files..."
+	find . -name '*.go' | while read -r file; do gofmt -w -s "$$file"; goimports -w "$$file"; done
+
+.PHONY: test
+test: ## Run tests
```
Comment (edaniels) [inline Makefile:14]: Seems easier to just run from your command line `go test -race ./...` and add coverage if you want.

File: Makefile [hunk 3]
```diff
@@ -0,0 +1,25 @@
+# A Self-Documenting Makefile: http://marmelab.com/blog/2016/02/29/auto-documented-makefile.html
```
Comment (edaniels) [inline Makefile:1]: I'd rather not introduce a Makefile to such a small project

File: digest.go [hunk 4]
```diff
@@ -139,28 +141,34 @@ type credentials struct {
 	NonceCount int
 	method     string
 	password   string
+	impl       hashing
 }
 
-func h(data string) string {
-	hf := md5.New()
+type hashing func() hash.Hash
```
Comment (edaniels) [inline digest.go:147]: [nit] hashingFUnc

File: digest.go [hunk 5]
```diff
@@ -224,6 +232,12 @@ func (t *Transport) newCredentials(req *http.Request, c *challenge) *credentials
 		method:     req.Method,
 		password:   t.Password,
 	}
+	if c.Algorithm == "MD5" {
```
Comment (edaniels) [inline digest.go:235]: we should probably error here earlier on if we're going to put the struct in a bad state

File: digest.go [hunk 6]
```diff
@@ -139,28 +141,34 @@ type credentials struct {
...
 }
 
 func (c *credentials) ha2() string {
-	return h(fmt.Sprintf("%s:%s", c.method, c.DigestURI))
+	return h(fmt.Sprintf("%s:%s", c.method, c.DigestURI), c.impl)
 }
 
 func (c *credentials) resp(cnonce string) (string, error) {
 	c.NonceCount++
+	if c.impl == nil {
```
Comment (edaniels) [inline digest.go:169]: better to error earlier at credentials construction than here

File: digest_test.go [hunk 7]
```diff
@@ -22,69 +22,194 @@
...
-}
-
 var cnonce = "0a4f113b"
 
 func TestH(t *testing.T) {
-	r1 := h("Mufasa:testrealm@host.com:Circle Of Life")
-	if r1 != "939e7578ed9e3c518a452acee763bce9" {
-		t.Fail()
-	}
+	t.Run("MD5", func(t *testing.T) {
```
Comment (edaniels) [inline digest_test.go:34]: can we make this a table based test instead to reduce the amount of code?

File: digest_test.go [hunk 8]
```diff
@@ -22,69 +22,194 @@
...
+	})
 }
 
 func TestKd(t *testing.T) {
-	r1 := kd("939e7578ed9e3c518a452acee763bce9",
-		"dcd98b7102dd2f0e8b11d0f600bfb0c093:00000001:0a4f113b:auth:39aff3a2bab6126f332b942af96d3366")
-	if r1 != "6629fae49393a05397450978507c4ef1" {
-		t.Fail()
-	}
+	t.Run("MD5", func(t *testing.T) {
```
Comment (edaniels) [inline digest_test.go:75]: same table based comments for each of these below

**Review by edaniels — CHANGES_REQUESTED** [review]
Done with a first pass!

**Comment by gssbzn** [comment]
replaced with #3

**Pull request closed without merging**